src/layer/vector/Canvas.js

```diff
@@ -354,7 +356,11 @@ export var Canvas = Renderer.extend({
 	// so we emulate that by calculating what's under the mouse on mousemove/click manually
 
 	_onClick: function (e) {
-		var point = this._map.mouseEventToLayerPoint(e), layer, clickedLayer;
+		var first = e.touches ? e.touches[0] : e;
+		if (['touchend', 'touchcancel', 'pointerup', 'pointercancel'].indexOf(e.type) > -1) {
+			first = e.changedTouches[0];
+		}
+		var point = this._map.mouseEventToLayerPoint(first), layer, clickedLayer;
```

Comment:
Instead of doing this for every `mouseEventToLayerPoint` call, we can update the `getMousePosition`:
https://github.com/Leaflet/Leaflet/blob/e3719a317ab325ed091fbbc6bfd34640fe19e13e/src/dom/DomEvent.js#L230-L244